src/shared/lending/FortressLendingCore.sol

```diff
@@ -0,0 +1,928 @@
+            _amount = (_shares * _totalAmount) / _totalSupply;
+            if (_roundUp && (_amount * _totalSupply) / _totalAmount < _shares) {
+                _amount = _amount + 1;
+            }
+        }
+    }
+
+    /// @notice Allows an on-chain or off-chain user to simulate the effects of their redeemption at the current block, given current on-chain conditions
+    /// @param _shares - The amount of _shares to redeem
+    /// @return - The amount of _assets in return
+    function previewRedeem(uint256 _shares) public view override returns (uint256) {
+        uint256 _assets = convertToAssets(_shares);
+
+        if (_assets > _totalAssetAvailable()) revert InsufficientAssetsInContract(_assets, _totalAssetAvailable());
+
+        return _assets;
+    }
+
+    /// @notice Allows an on-chain or off-chain user to simulate the effects of their withdrawal at the current block, given current on-chain conditions
+    /// @param _assets - The amount of _assets to withdraw
+    /// @return - The amount of shares to burn
+    function previewWithdraw(uint256 _assets) public view override returns (uint256) {
+        if (_assets > _totalAssetAvailable()) revert InsufficientAssetsInContract(_assets, _totalAssetAvailable());
+
+        return convertToShares(_assets);
+    }
+
+    /// @notice Returns the values of all constants used in the contract
+    function getConstants() external pure
+        returns (
+            uint256 _LTV_PRECISION,
+            uint256 _LIQ_PRECISION,
+            uint256 _UTIL_PREC,
+            uint256 _FEE_PRECISION,
+            uint256 _EXCHANGE_PRECISION,
+            uint64 _DEFAULT_INT,
+            uint16 _DEFAULT_PROTOCOL_FEE,
+            uint256 _MAX_PROTOCOL_FEE
+        )
+    {
+        _LTV_PRECISION = LTV_PRECISION;
+        _LIQ_PRECISION = LIQ_PRECISION;
+        _UTIL_PREC = UTIL_PREC;
+        _FEE_PRECISION = FEE_PRECISION;
+        _EXCHANGE_PRECISION = EXCHANGE_PRECISION;
+        _DEFAULT_INT = DEFAULT_INT;
+        _DEFAULT_PROTOCOL_FEE = DEFAULT_PROTOCOL_FEE;
+        _MAX_PROTOCOL_FEE = MAX_PROTOCOL_FEE;
+    }
+
+    // ============================================================================================
+    // Internal Helpers
+    // ============================================================================================
+
+    /// @notice Returns the total balance of Asset Tokens in the contract
+    /// @return The balance of Asset Tokens held by contract
+    function _totalAssetAvailable() internal view returns (uint256) {
+        return totalAssets() - totalBorrow.amount;
+    }
+
+    /// @notice Determines if a given borrower is solvent given an exchange rate
+    /// @param _borrower The borrower address to check
+    /// @param _exchangeRate The exchange rate, i.e. the amount of collateral to buy 1e18 asset
+    /// @return Whether borrower is solvent
+    function _isSolvent(address _borrower, uint256 _exchangeRate) internal view returns (bool) {
```

Comment:
I am reading this function as a generic solvency check for a given address. Why would any exchange rate be given as parameter here? Doesn't that open up certain attack vectors? Maybe this could be more resilient by verifying solvency given a specific position held by the given address?

Reply:
It's an `internal` `read` function. No user input ever gets to that function.